File: module/dependency.py
import logging, os
from pathlib import Path
from datetime import datetime

from etc import constants
from . import properties
from . import dict_tools
from . import build_cmds
from . import files


default_app_config = properties.get_app_properties()


def parse_dependency_file(file_path):

  return properties.get_config(file_path)



def get_build_order(dependency_dict, target_list=[], app_config=properties.get_app_properties()):

  objects_tree = get_targets_depended_objects(dependency_dict, target_list)
  files.writeJson(objects_tree, '.obi/tmp/objects_tree.json')

  dependend_objects = get_targets_only_depended_objects(dependency_dict, target_list)
  logging.debug(f"{objects_tree=}")
  files.writeJson(dependend_objects, constants.DEPENDEND_OBJECT_LIST)

  ordered_target_tree = get_targets_by_level(objects_tree)
  logging.debug(f"{ordered_target_tree=}")
  files.writeJson(ordered_target_tree, '.obi/tmp/ordered_target_tree.json')

  new_target_tree = remove_duplicities(ordered_target_tree)
  files.writeJson(new_target_tree, '.obi/tmp/new_target_tree.json')
  logging.debug(f"{new_target_tree=}")

  build_cmds.add_build_cmds(new_target_tree, app_config)

  new_target_tree = {'timestamp': str(datetime.now()), 'compiles': new_target_tree}

  return new_target_tree




def get_all_targets(dependency_dict, target_list=[]):
  tree = target_list

  for obj, obj_dependencies in dependency_dict.items():
    if any(item in target_list for item in obj_dependencies):
      #tree.append(obj) #build_dependency_tree(dependency_dict, dep))
      tree.extend(build_dependency_tree(dependency_dict, [obj]))

  return tree



def get_targets_depended_objects(dependency_dict, targets=[], result={}):
  """Get sources as dict with it's dependend sources
    {'source1': {'source3', 'source4'}, 'source2': {}}

  Args:
      dependency_dict (_type_): _description_
      targets (list, optional): _description_. Defaults to [].
      result (dict, optional): _description_. Defaults to {}.

  Returns:
      _type_: _description_
  """

  targets_objects = {}

  for target in targets:

    targets_objects[target] = get_target_depended_objects(dependency_dict, target)

  return targets_objects



def get_target_depended_objects(dependency_dict, target, result={}):

  dependend_objects = {}
  src_base_path = default_app_config['general'].get('source-dir', 'src')

  for obj, obj_dependencies in dependency_dict.items():
    if target in obj_dependencies:

      if not Path(os.path.join(src_base_path, obj)).is_file():
        logging.debug(f"Doesn't exist: {obj=}, {Path(obj).is_file()=}")
        continue
    
      logging.debug(f"Add obj {obj=}")
      #tree.append(obj) #build_dependency_tree(dependency_dict, dep))
      dependend_objects[obj]=get_target_depended_objects(dependency_dict, obj, result)
  
  return dependend_objects



def get_targets_only_depended_objects(dependency_dict, targets=[]) -> []:
  """Get dependencies of the targets

  Args:
      dependency_dict (dict}): All dependencies of all sources
      targets (list, optional): List of targets. Defaults to []

  Returns:
      list: Returns a list with dependencies
  """

  targets_objects = []

  for target in targets:
    logging.debug(f"Dependend 1 {target}")
    targets_objects = targets_objects + get_target_only_depended_objects(dependency_dict, target, targets)

  return list(set(targets_objects))



def get_target_only_depended_objects(dependency_dict, target, orig_targets:[]):

  dependend_objects = []
  src_base_path = default_app_config['general'].get('source-dir', 'src')

  for obj, obj_dependencies in dependency_dict.items():
    if target in obj_dependencies:
      
      if not Path(os.path.join(src_base_path, obj)).is_file():
        logging.warning(f"Doesn't exist: {obj=}, {Path(obj).is_file()=}")
        continue
      
      if (obj in orig_targets):
        continue

      #tree.append(obj) #build_dependency_tree(dependency_dict, dep))
      dependend_objects = dependend_objects + [obj] + get_target_only_depended_objects(dependency_dict, obj, orig_targets)
  
  return dependend_objects



def remove_duplicities(target_tree=[]):

  # All levels
  for level_item in sorted(target_tree, key=lambda d: d['level']):

    # All objects in this level
    for obj in level_item['sources']:

      # Scan reverse for duplicated objects
      for rev_level_item in reversed(sorted(target_tree, key=lambda d: d['level'])):

        rev_level_sources = [item['source'] for item in rev_level_item['sources']]

        for i in range(len(target_tree)):
          sources = [item['source'] for item in target_tree[i]['sources']]
          if target_tree[i]['level'] < rev_level_item['level'] and obj['source'] in rev_level_sources and obj['source'] in sources:
            target_tree[i]['sources'].remove(obj)

  return target_tree




def get_targets_by_level(target_tree={}, level=1):
  
  new_target_tree = []
  
  # Add object to list
  for obj, next_objs in target_tree.items():

    loop_level_obj = {}
    loop_level_obj = [item for item in new_target_tree if item['level'] == level]
    if len(loop_level_obj) > 0:
      loop_level_obj = loop_level_obj[0]

    #if level not in new_target_tree.keys():
    if not loop_level_obj:
      loop_level_obj = {'level': level, 'sources': []}
      new_target_tree.append(loop_level_obj)

    #if obj in new_target_tree[level]:
    if obj in [item['source'] for item in loop_level_obj['sources']]:
      continue

    loop_level_obj['sources'].append({'source': obj, 'cmds': []})

    # Also add dependend objects to list
    for next_obj, next_sub_objs in next_objs.items():

      loop_next_level_obj = {}
      loop_next_level_obj = [item for item in new_target_tree if item['level'] == level+1]
      if loop_next_level_obj:
        loop_next_level_obj = loop_next_level_obj[0]

      if not loop_next_level_obj:
        loop_next_level_obj = {'level': level+1, 'sources': []}
        new_target_tree.append(loop_next_level_obj)
      
      if next_obj in [item['source'] for item in loop_next_level_obj['sources']]:
        continue

      loop_next_level_obj['sources'].append({'source': next_obj, 'cmds': []})

      # Recursive call to go through the tree
      extended_tree = get_targets_by_level(next_sub_objs, level+2)
      new_target_tree = dict_tools.deep_list_merge(extended_tree, new_target_tree)

  return sorted(new_target_tree, key=lambda d: d['level'])

File: src/module/dependency.ts
// TypeScript conversion of module/dependency.py

import * as fs from 'fs';
import * as path from 'path';
import * as properties from './properties';
import * as dictTools from './dict_tools';
import * as buildCmds from './build_cmds';
import * as files from './files';
import { OBIConstants } from './obi_constants';

interface SourceItem {
    source: string;
    cmds: any[];
}

interface LevelItem {
    level: number;
    sources: SourceItem[];
}

interface DependencyDict {
    [key: string]: string[];
}

interface ObjectsTree {
    [key: string]: ObjectsTree;
}

const defaultAppConfig = properties.getAppProperties();

export function parseDependencyFile(filePath: string): DependencyDict {
    return properties.getJson(filePath);
}

export function getBuildOrder(
    dependencyDict: DependencyDict,
    targetList: string[] = [],
    appConfig: any = defaultAppConfig
): { timestamp: string; compiles: LevelItem[] } {
    const objectsTree = getTargetsDependedObjects(dependencyDict, targetList);
    files.writeJson(objectsTree, '.obi/tmp/objects_tree.json');

    const dependedObjects = getTargetsOnlyDependedObjects(dependencyDict, targetList);
    console.log(`objectsTree=${JSON.stringify(objectsTree)}`);
    files.writeJson(dependedObjects, OBIConstants.get("DEPENDEND_OBJECT_LIST") || '.obi/tmp/dependend-object-list.json');

    const orderedTargetTree = getTargetsByLevel(objectsTree);
    console.log(`orderedTargetTree=${JSON.stringify(orderedTargetTree)}`);
    files.writeJson(orderedTargetTree, '.obi/tmp/ordered_target_tree.json');

    const newTargetTree = removeDuplicities(orderedTargetTree);
    files.writeJson(newTargetTree, '.obi/tmp/new_target_tree.json');
    console.log(`newTargetTree=${JSON.stringify(newTargetTree)}`);

    buildCmds.addBuildCmds(newTargetTree, appConfig);

    const result = {
        timestamp: new Date().toISOString(),
        compiles: newTargetTree
    };

    return result;
}

function getTargetsDependedObjects(
    dependencyDict: DependencyDict,
    targets: string[] = []
): ObjectsTree {
    const targetsObjects: ObjectsTree = {};

    for (const target of targets) {
        targetsObjects[target] = getTargetDependedObjects(dependencyDict, target);
    }

    return targetsObjects;
}

function getTargetDependedObjects(
    dependencyDict: DependencyDict,
    target: string,
    result: ObjectsTree = {}
): ObjectsTree {
    const dependedObjects: ObjectsTree = {};
    const srcBasePath = defaultAppConfig['general']?.['source-dir'] || 'src';

    for (const [obj, objDependencies] of Object.entries(dependencyDict)) {
        if (objDependencies.includes(target)) {
            const objPath = path.join(srcBasePath, obj);
            if (!fs.existsSync(objPath)) {
                console.log(`Doesn't exist: obj=${obj}, exists=${fs.existsSync(objPath)}`);
                continue;
            }
            
            console.log(`Add obj obj=${obj}`);
            dependedObjects[obj] = getTargetDependedObjects(dependencyDict, obj, result);
        }
    }
    
    return dependedObjects;
}

function getTargetsOnlyDependedObjects(
    dependencyDict: DependencyDict,
    targets: string[] = []
): string[] {
    let targetsObjects: string[] = [];

    for (const target of targets) {
        console.log(`Dependend 1 ${target}`);
        targetsObjects = targetsObjects.concat(getTargetOnlyDependedObjects(dependencyDict, target, targets));
    }

    return Array.from(new Set(targetsObjects));
}

function getTargetOnlyDependedObjects(
    dependencyDict: DependencyDict,
    target: string,
    origTargets: string[]
): string[] {
    let dependedObjects: string[] = [];
    const srcBasePath = defaultAppConfig['general']?.['source-dir'] || 'src';

    for (const [obj, objDependencies] of Object.entries(dependencyDict)) {
        if (objDependencies.includes(target)) {
            const objPath = path.join(srcBasePath, obj);
            if (!fs.existsSync(objPath)) {
                console.warn(`Doesn't exist: obj=${obj}, exists=${fs.existsSync(objPath)}`);
                continue;
            }
            
            if (origTargets.includes(obj)) {
                continue;
            }

            dependedObjects = dependedObjects.concat([obj], getTargetOnlyDependedObjects(dependencyDict, obj, origTargets));
        }
    }
    
    return dependedObjects;
}

function removeDuplicities(targetTree: LevelItem[]): LevelItem[] {
    // All levels
    for (const levelItem of targetTree.sort((a, b) => a.level - b.level)) {
        // All objects in this level
        for (const obj of levelItem.sources) {
            // Scan reverse for duplicated objects
            for (const revLevelItem of targetTree.slice().reverse().sort((a, b) => b.level - a.level)) {
                const revLevelSources = revLevelItem.sources.map(item => item.source);

                for (let i = 0; i < targetTree.length; i++) {
                    const sources = targetTree[i].sources.map(item => item.source);
                    if (targetTree[i].level < revLevelItem.level && 
                        revLevelSources.includes(obj.source) && 
                        sources.includes(obj.source)) {
                        const index = targetTree[i].sources.findIndex(item => item.source === obj.source);
                        if (index !== -1) {
                            targetTree[i].sources.splice(index, 1);
                        }
                    }
                }
            }
        }
    }

    return targetTree;
}

function getTargetsByLevel(targetTree: ObjectsTree, level: number = 1): LevelItem[] {
    let newTargetTree: LevelItem[] = [];
    
    // Add object to list
    for (const [obj, nextObjs] of Object.entries(targetTree)) {
        let loopLevelObj = newTargetTree.find(item => item.level === level);
        
        if (!loopLevelObj) {
            loopLevelObj = { level, sources: [] };
            newTargetTree.push(loopLevelObj);
        }

        if (loopLevelObj.sources.some(item => item.source === obj)) {
            continue;
        }

        loopLevelObj.sources.push({ source: obj, cmds: [] });

        // Also add depended objects to list
        for (const [nextObj, nextSubObjs] of Object.entries(nextObjs as ObjectsTree)) {
            let loopNextLevelObj = newTargetTree.find(item => item.level === level + 1);
            
            if (!loopNextLevelObj) {
                loopNextLevelObj = { level: level + 1, sources: [] };
                newTargetTree.push(loopNextLevelObj);
            }
            
            if (loopNextLevelObj.sources.some(item => item.source === nextObj)) {
                continue;
            }

            loopNextLevelObj.sources.push({ source: nextObj, cmds: [] });

            // Recursive call to go through the tree
            const extendedTree = getTargetsByLevel(nextSubObjs as ObjectsTree, level + 2);
            newTargetTree = dictTools.deepListMerge(extendedTree, newTargetTree);
        }
    }

    return newTargetTree.sort((a, b) => a.level - b.level);
}
